// @remotion/studio/dist/esm/chunk-jefhcs5z.js
var __create = Object.create;
var __getProtoOf = Object.getPrototypeOf;
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __toESM = (mod, isNodeMode, target) => {
  target = mod != null ? __create(__getProtoOf(mod)) : {};
  const to = isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target;
  for (let key of __getOwnPropNames(mod))
    if (!__hasOwnProp.call(to, key))
      __defProp(to, key, {
        get: () => mod[key],
        enumerable: true
      });
  return to;
};
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/helpers/colors.ts
var BLACK = "black";
var WHITE = "white";
var RED = "red";
var TRANSPARENT = "transparent";
var CURRENT_COLOR = "currentColor";
var CURRENT_COLOR_LOWERCASE = "currentcolor";
var WHITE_HEX = "#fff";
var WHITE_FULL_HEX = "#ffffff";
var BLACK_HEX = "#000";
var BLACK_FULL_HEX = "#000000";
var BACKGROUND = "rgb(31,36,40)";
var BACKGROUND_HEX = "#1f2428";
var BACKGROUND__TRANSPARENT = "rgba(31,36,40, 0)";
var INPUT_BACKGROUND = "#2f363d";
var LIGHT_COLOR = "#ddd";
var SELECTED_BACKGROUND = "hsla(0, 0%, 100%, 0.15)";
var LIGHT_TEXT = "#A6A7A9";
var RULER_COLOR = "#808080";
var WHITE_ALPHA_05 = "rgba(255, 255, 255, 0.05)";
var WHITE_ALPHA_06 = "rgba(255, 255, 255, 0.06)";
var WHITE_ALPHA_08 = "rgba(255, 255, 255, 0.08)";
var WHITE_ALPHA_10 = "rgba(255, 255, 255, 0.1)";
var WHITE_ALPHA_12 = "rgba(255, 255, 255, 0.12)";
var WHITE_ALPHA_15 = "rgba(255, 255, 255, 0.15)";
var WHITE_ALPHA_20 = "rgba(255, 255, 255, 0.2)";
var WHITE_ALPHA_25 = "rgba(255, 255, 255, 0.25)";
var WHITE_ALPHA_30 = "rgba(255, 255, 255, 0.3)";
var WHITE_ALPHA_35 = "rgba(255, 255, 255, 0.35)";
var WHITE_ALPHA_40 = "rgba(255, 255, 255, 0.4)";
var WHITE_ALPHA_45 = "rgba(255, 255, 255, 0.45)";
var WHITE_ALPHA_50 = "rgba(255, 255, 255, 0.5)";
var WHITE_ALPHA_60 = "rgba(255, 255, 255, 0.6)";
var WHITE_ALPHA_70 = "rgba(255, 255, 255, 0.7)";
var WHITE_ALPHA_72 = "rgba(255, 255, 255, 0.72)";
var WHITE_ALPHA_80 = "rgba(255, 255, 255, 0.8)";
var BLACK_ALPHA_10 = "rgba(0, 0, 0, 0.1)";
var BLACK_ALPHA_28 = "rgba(0, 0, 0, 0.28)";
var BLACK_ALPHA_30 = "rgba(0, 0, 0, 0.3)";
var BLACK_ALPHA_40 = "rgba(0, 0, 0, 0.4)";
var BLACK_ALPHA_50 = "rgba(0, 0, 0, 0.5)";
var BLACK_ALPHA_60 = "rgba(0, 0, 0, 0.6)";
var BLACK_ALPHA_80 = "rgba(0, 0, 0, 0.8)";
var BLACK_ALPHA_85 = "rgba(0, 0, 0, 0.85)";
var BLACK_ALPHA_90 = "rgba(0, 0, 0, 0.9)";
var BLACK_OPAQUE = "rgba(0, 0, 0, 1)";
var SELECTED_HOVER_BACKGROUND = "hsla(0, 0%, 100%, 0.25)";
var FAIL_COLOR = "#ff3232";
var WARNING_COLOR = "#f1c40f";
var BLUE = "#0b84f3";
var BLUE_HOVERED = "#4da3f7";
var BLUE_DISABLED = "#284f73";
var UNSELECTED_GUIDE = "#7e1219";
var SELECTED_GUIDE = "#d22d3a";
var LINE_COLOR = "#363A3E";
var TIMELINE_TRACK_SEPARATOR = "#13161B";
var LIGHT_GRAY = "#ccc";
var KEYBOARD_SHORTCUT_KEY_COLOR = "#eee";
var NOTIFICATION_BACKGROUND = "#111111";
var RENDER_STATUS_BACKGROUND = "#222";
var ERROR_CODE_FRAME_BACKGROUND = "#070707";
var ERROR_CODE_FRAME_LINE_BACKGROUND = "#121212";
var ERROR_LINK_COLOR = "#58a6ff";
var INFO_BLUE = "#60a5fa";
var SERVER_DISCONNECTED_BACKGROUND = "#e74c3c";
var STACK_FRAME_BORDER_BLUE = "rgb(66, 144, 245)";
var TIMELINE_BACKGROUND_COLOR = "#0F1113";
var TIMELINE_SELECTED_BACKGROUND_COLOR = "#3B3F42";
var TIMELINE_SELECTED_LABEL_BACKGROUND_COLOR = "#B0B0B0";
var TIMELINE_BLUE = "#0b84ff";
var TIMELINE_PLAYHEAD_COLOR = "#f02c00";
var TIMELINE_DROP_BLUE_ALPHA_12 = "rgba(0, 155, 255, 0.12)";
var TIMELINE_DROP_BLUE_ALPHA_16 = "rgba(0, 155, 255, 0.16)";
var TIMELINE_DROP_BLUE_ALPHA_75 = "rgba(0, 155, 255, 0.75)";
var TIMELINE_MARQUEE_BLUE_ALPHA_16 = "rgba(70, 130, 255, 0.16)";
var TIMELINE_MARQUEE_BLUE_ALPHA_75 = "rgba(70, 130, 255, 0.75)";
var EASING_SELECTED_BACKGROUND = "rgba(11, 132, 243, 0.18)";
var INFO_BLUE_BACKGROUND = "rgba(59, 130, 246, 0.15)";
var BORDER_BLACK = `1px solid ${BLACK_HEX}`;
var BORDER_WHITE = `1px solid ${WHITE_HEX}`;
var BORDER_WHITE_2PX = `2px solid ${WHITE_HEX}`;
var BORDER_CURRENT_COLOR = `1px solid ${CURRENT_COLOR}`;
var BORDER_TRANSPARENT_2PX = `2px solid ${TRANSPARENT}`;
var BORDER_BLACK_ALPHA_50 = `1px solid ${BLACK_ALPHA_50}`;
var BORDER_BLACK_ALPHA_60 = `1px solid ${BLACK_ALPHA_60}`;
var BORDER_WHITE_ALPHA_12 = `1px solid ${WHITE_ALPHA_12}`;
var BORDER_WHITE_ALPHA_20 = `1px solid ${WHITE_ALPHA_20}`;
var BORDER_INFO_BLUE = "1px solid rgba(59, 130, 246, 0.4)";
var BORDER_STACK_FRAME_BLUE = `1px solid ${STACK_FRAME_BORDER_BLUE}`;
var BORDER_TIMELINE_DROP_BLUE = `1px solid ${TIMELINE_DROP_BLUE_ALPHA_75}`;
var BORDER_TIMELINE_MARQUEE_BLUE = `1px solid ${TIMELINE_MARQUEE_BLUE_ALPHA_75}`;
var SHADOW_BLACK = `0 0 4px ${BLACK}`;
var SHADOW_BLACK_ALPHA_50_TOWARDS_BOTTOM = `0 2px 8px ${BLACK_ALPHA_50}`;
var SHADOW_BLACK_ALPHA_50_TOWARDS_TOP = `0 -2px 8px ${BLACK_ALPHA_50}`;
var NOTIFICATION_SHADOW = `0 2px 3px ${BLACK_OPAQUE}`;
var SERVER_DISCONNECTED_SHADOW = `0 2px 4px ${BLACK_ALPHA_40}`;
var COLOR_PICKER_POPUP_SHADOW = `0 4px 16px ${BLACK_ALPHA_50}`;
var COLOR_PICKER_HANDLE_SHADOW = `0 0 0 1px ${BLACK_ALPHA_60}`;
var NOTIFICATION_BORDER = `0.25px solid ${WHITE_ALPHA_10}`;
var FOCUS_BOX_SHADOW = `inset 1px 1px #555, inset -1px -1px #555, ` + `inset 1px -1px #555, inset -1px 1px #555`;
var CHECKERBOARD_BACKGROUND_IMAGE = `
     linear-gradient(
        45deg,
        ${BLACK_ALPHA_10} 25%,
        ${TRANSPARENT} 25%
      ),
      linear-gradient(135deg, ${BLACK_ALPHA_10} 25%, ${TRANSPARENT} 25%),
      linear-gradient(45deg, ${TRANSPARENT} 75%, ${BLACK_ALPHA_10} 75%),
      linear-gradient(135deg, ${TRANSPARENT} 75%, ${BLACK_ALPHA_10} 75%)
    `;
var COLOR_PICKER_CHECKER_BACKGROUND_IMAGE = `linear-gradient(45deg, #888 25%, ${TRANSPARENT} 25%), ` + `linear-gradient(-45deg, #888 25%, ${TRANSPARENT} 25%), ` + `linear-gradient(45deg, ${TRANSPARENT} 75%, #888 75%), ` + `linear-gradient(-45deg, ${TRANSPARENT} 75%, #888 75%)`;
var COLOR_PICKER_CHECKER_BACKGROUND_COLOR = "#444";
var COLOR_PICKER_SATURATION_VALUE_GRADIENT = `linear-gradient(to right, ${WHITE_HEX}, ${TRANSPARENT})`;
var COLOR_PICKER_SATURATION_BLACK_GRADIENT = `linear-gradient(to top, ${BLACK_HEX}, ${TRANSPARENT})`;
var COLOR_PICKER_ALPHA_TRANSPARENT = "rgba(0, 0, 0, 0)";
var COLOR_PICKER_HUE_GRADIENT = "linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)";
var ERROR_MESSAGE_MASK_IMAGE = `linear-gradient(to bottom, ${WHITE} 60%, ${TRANSPARENT})`;
var SELECTED_OUTLINE_DROP_SHADOW = `drop-shadow(0 0 1px ${WHITE_ALPHA_20})`;
var SELECTED_OUTLINE_UV_DROP_SHADOW = `drop-shadow(0 1px 2px ${BLACK_ALPHA_28})`;
var LOOPED_INDICATOR_DROP_SHADOW = `drop-shadow(0 0 2px ${BLACK_ALPHA_90}) ` + `drop-shadow(0 1px 2px ${BLACK_ALPHA_80})`;
var TIMELINE_AUDIO_GRADIENT = "linear-gradient(rgb(16 171 58), rgb(43 165 63) 60%)";
var TIMELINE_VIDEO_GRADIENT = "linear-gradient(to top, #8e44ad, #9b59b6)";
var TIMELINE_IMAGE_GRADIENT = "linear-gradient(to top, #2980b9, #3498db)";
var getBackgroundFromHoverState = ({
  selected,
  hovered
}) => {
  if (selected) {
    if (hovered) {
      return SELECTED_HOVER_BACKGROUND;
    }
    return SELECTED_BACKGROUND;
  }
  if (hovered) {
    return WHITE_ALPHA_06;
  }
  return TRANSPARENT;
};

export { __toESM, __require, BLACK, WHITE, RED, TRANSPARENT, CURRENT_COLOR, CURRENT_COLOR_LOWERCASE, WHITE_HEX, WHITE_FULL_HEX, BLACK_HEX, BLACK_FULL_HEX, BACKGROUND, BACKGROUND_HEX, BACKGROUND__TRANSPARENT, INPUT_BACKGROUND, LIGHT_COLOR, SELECTED_BACKGROUND, LIGHT_TEXT, RULER_COLOR, WHITE_ALPHA_05, WHITE_ALPHA_06, WHITE_ALPHA_08, WHITE_ALPHA_10, WHITE_ALPHA_12, WHITE_ALPHA_15, WHITE_ALPHA_20, WHITE_ALPHA_25, WHITE_ALPHA_30, WHITE_ALPHA_35, WHITE_ALPHA_40, WHITE_ALPHA_45, WHITE_ALPHA_50, WHITE_ALPHA_60, WHITE_ALPHA_70, WHITE_ALPHA_72, WHITE_ALPHA_80, BLACK_ALPHA_30, BLACK_ALPHA_40, BLACK_ALPHA_50, BLACK_ALPHA_60, BLACK_ALPHA_85, FAIL_COLOR, WARNING_COLOR, BLUE, BLUE_HOVERED, BLUE_DISABLED, UNSELECTED_GUIDE, SELECTED_GUIDE, LINE_COLOR, TIMELINE_TRACK_SEPARATOR, LIGHT_GRAY, KEYBOARD_SHORTCUT_KEY_COLOR, NOTIFICATION_BACKGROUND, RENDER_STATUS_BACKGROUND, ERROR_CODE_FRAME_BACKGROUND, ERROR_CODE_FRAME_LINE_BACKGROUND, ERROR_LINK_COLOR, INFO_BLUE, SERVER_DISCONNECTED_BACKGROUND, TIMELINE_BACKGROUND_COLOR, TIMELINE_SELECTED_BACKGROUND_COLOR, TIMELINE_SELECTED_LABEL_BACKGROUND_COLOR, TIMELINE_BLUE, TIMELINE_PLAYHEAD_COLOR, TIMELINE_DROP_BLUE_ALPHA_12, TIMELINE_DROP_BLUE_ALPHA_16, TIMELINE_MARQUEE_BLUE_ALPHA_16, EASING_SELECTED_BACKGROUND, INFO_BLUE_BACKGROUND, BORDER_BLACK, BORDER_WHITE, BORDER_WHITE_2PX, BORDER_CURRENT_COLOR, BORDER_TRANSPARENT_2PX, BORDER_BLACK_ALPHA_50, BORDER_BLACK_ALPHA_60, BORDER_WHITE_ALPHA_12, BORDER_WHITE_ALPHA_20, BORDER_INFO_BLUE, BORDER_STACK_FRAME_BLUE, BORDER_TIMELINE_DROP_BLUE, BORDER_TIMELINE_MARQUEE_BLUE, SHADOW_BLACK, SHADOW_BLACK_ALPHA_50_TOWARDS_BOTTOM, SHADOW_BLACK_ALPHA_50_TOWARDS_TOP, NOTIFICATION_SHADOW, SERVER_DISCONNECTED_SHADOW, COLOR_PICKER_POPUP_SHADOW, COLOR_PICKER_HANDLE_SHADOW, NOTIFICATION_BORDER, FOCUS_BOX_SHADOW, CHECKERBOARD_BACKGROUND_IMAGE, COLOR_PICKER_CHECKER_BACKGROUND_IMAGE, COLOR_PICKER_CHECKER_BACKGROUND_COLOR, COLOR_PICKER_SATURATION_VALUE_GRADIENT, COLOR_PICKER_SATURATION_BLACK_GRADIENT, COLOR_PICKER_ALPHA_TRANSPARENT, COLOR_PICKER_HUE_GRADIENT, ERROR_MESSAGE_MASK_IMAGE, SELECTED_OUTLINE_DROP_SHADOW, SELECTED_OUTLINE_UV_DROP_SHADOW, LOOPED_INDICATOR_DROP_SHADOW, TIMELINE_AUDIO_GRADIENT, TIMELINE_VIDEO_GRADIENT, TIMELINE_IMAGE_GRADIENT, getBackgroundFromHoverState };
