// @remotion/studio/dist/components/Timeline/save-sequence-prop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveSequenceProps = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const call_api_1 = require("../call-api");
const call_move_keyframe_1 = require("./call-move-keyframe");
const save_prop_queue_1 = require("./save-prop-queue");
const serializeSequencePropValue = (value) => {
    if (value === undefined) {
        return { type: 'undefined' };
    }
    return { type: 'json', serialized: JSON.stringify(value) };
};
const saveSequenceProps = ({ changes, movedKeyframes, setPropStatuses, clientId, undoLabel, redoLabel, }) => {
    var _a, _b;
    const sequenceKeyframes = (_a = movedKeyframes === null || movedKeyframes === void 0 ? void 0 : movedKeyframes.sequenceKeyframes) !== null && _a !== void 0 ? _a : [];
    const effectKeyframes = (_b = movedKeyframes === null || movedKeyframes === void 0 ? void 0 : movedKeyframes.effectKeyframes) !== null && _b !== void 0 ? _b : [];
    if (changes.length === 0 &&
        sequenceKeyframes.length === 0 &&
        effectKeyframes.length === 0) {
        return Promise.resolve();
    }
    if (changes.length === 1 &&
        sequenceKeyframes.length === 0 &&
        effectKeyframes.length === 0) {
        const change = changes[0];
        if (change === undefined) {
            throw new Error('Expected a sequence prop change');
        }
        return (0, save_prop_queue_1.enqueueSavePropChange)({
            nodePath: change.nodePath,
            setPropStatuses,
            applyOptimistic: (prev) => (0, studio_shared_1.optimisticUpdateForPropStatuses)({
                previous: prev,
                fieldKey: change.fieldKey,
                value: change.value,
                defaultValue: change.defaultValue,
                schema: change.schema,
            }),
            apiCall: () => {
                var _a;
                return (0, call_api_1.callApi)('/api/save-sequence-props', {
                    edits: [
                        {
                            fileName: change.fileName,
                            nodePath: change.nodePath,
                            key: change.fieldKey,
                            value: serializeSequencePropValue(change.value),
                            defaultValue: change.defaultValue,
                            schema: change.schema,
                            sourceEdit: (_a = change.sourceEdit) !== null && _a !== void 0 ? _a : null,
                        },
                    ],
                    clientId,
                    undoLabel,
                    redoLabel,
                });
            },
            errorLabel: 'Could not save sequence prop',
        });
    }
    (0, call_move_keyframe_1.applyOptimisticKeyframeMoves)({
        sequenceKeyframes,
        effectKeyframes,
        setPropStatuses,
    });
    for (const change of changes) {
        setPropStatuses(change.nodePath, (prev) => (0, studio_shared_1.optimisticUpdateForPropStatuses)({
            previous: prev,
            fieldKey: change.fieldKey,
            value: change.value,
            defaultValue: change.defaultValue,
            schema: change.schema,
        }));
    }
    return (0, call_api_1.callApi)('/api/save-sequence-props', {
        edits: changes.map((change) => {
            var _a;
            return {
                fileName: change.fileName,
                nodePath: change.nodePath,
                key: change.fieldKey,
                value: serializeSequencePropValue(change.value),
                defaultValue: change.defaultValue,
                schema: change.schema,
                sourceEdit: (_a = change.sourceEdit) !== null && _a !== void 0 ? _a : null,
            };
        }),
        movedKeyframes: {
            sequenceKeyframes: sequenceKeyframes.map((keyframe) => ({
                fileName: keyframe.fileName,
                nodePath: keyframe.nodePath,
                key: keyframe.fieldKey,
                fromFrame: keyframe.fromFrame,
                toFrame: keyframe.toFrame,
                schema: keyframe.schema,
            })),
            effectKeyframes: effectKeyframes.map((keyframe) => ({
                fileName: keyframe.fileName,
                sequenceNodePath: keyframe.nodePath,
                effectIndex: keyframe.effectIndex,
                key: keyframe.fieldKey,
                fromFrame: keyframe.fromFrame,
                toFrame: keyframe.toFrame,
                schema: keyframe.schema,
            })),
        },
        clientId,
        undoLabel,
        redoLabel,
    }).then(() => undefined);
};
exports.saveSequenceProps = saveSequenceProps;
