// @remotion/studio/dist/components/Timeline/split-selected-timeline-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitSelectedTimelineItems = exports.shouldHandleTimelineSplitShortcut = exports.shouldHandleTimelineDuplicateShortcut = exports.splitTimelineSequenceFromSource = exports.getTimelineSequenceSplitEligibility = exports.getTimelineSequenceSplitUnsupportedReason = void 0;
const remotion_1 = require("remotion");
const call_api_1 = require("../call-api");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const find_track_for_node_path_info_1 = require("./find-track-for-node-path-info");
const getTimelineSequenceSplitUnsupportedReason = (componentName) => {
    if (componentName === '<TransitionSeries.Sequence>' ||
        componentName === '<TransitionSeries.Overlay>') {
        return `${componentName} cannot be split from source`;
    }
    return null;
};
exports.getTimelineSequenceSplitUnsupportedReason = getTimelineSequenceSplitUnsupportedReason;
const staticNumberish = (status) => {
    if (!status) {
        return true;
    }
    return (status.status === 'static' &&
        (typeof status.codeValue === 'number' || status.codeValue === undefined));
};
const getTimelineSequenceSplitEligibility = ({ selection, sequence, splitFrame, propStatuses, }) => {
    var _a;
    if (selection.type !== 'sequence') {
        return {
            canSplit: false,
            reason: 'Select one sequence to split',
        };
    }
    if (!sequence) {
        return {
            canSplit: false,
            reason: 'Could not find selected sequence',
        };
    }
    if (!Number.isInteger(splitFrame)) {
        return {
            canSplit: false,
            reason: 'Split frame must be an integer',
        };
    }
    if (sequence.isInsideSeries) {
        return {
            canSplit: false,
            reason: 'Series.Sequence clips cannot be split from source',
        };
    }
    const unsupportedReason = (0, exports.getTimelineSequenceSplitUnsupportedReason)((_a = sequence.controls) === null || _a === void 0 ? void 0 : _a.componentName);
    if (unsupportedReason) {
        return {
            canSplit: false,
            reason: unsupportedReason,
        };
    }
    const { nodePathInfo } = selection;
    if (!nodePathInfo.sequenceSubscriptionKey.nodePath) {
        return {
            canSplit: false,
            reason: 'Sequence has no editable source node',
        };
    }
    if (nodePathInfo.numberOfSequencesWithThisNodePath > 1) {
        return {
            canSplit: false,
            reason: 'Programmatically duplicated sequences cannot be split',
        };
    }
    if (!staticNumberish(propStatuses === null || propStatuses === void 0 ? void 0 : propStatuses.from) ||
        !staticNumberish(propStatuses === null || propStatuses === void 0 ? void 0 : propStatuses.durationInFrames) ||
        !staticNumberish(propStatuses === null || propStatuses === void 0 ? void 0 : propStatuses.trimBefore)) {
        return {
            canSplit: false,
            reason: 'Sequence timing props must be static numbers',
        };
    }
    const start = sequence.from;
    const end = sequence.duration === Infinity
        ? Infinity
        : sequence.from + sequence.duration;
    if (splitFrame <= start) {
        return {
            canSplit: false,
            reason: 'Cannot split at the sequence start',
        };
    }
    if (splitFrame >= end) {
        return {
            canSplit: false,
            reason: 'Cannot split at the sequence end',
        };
    }
    return {
        canSplit: true,
        nodePathInfo,
    };
};
exports.getTimelineSequenceSplitEligibility = getTimelineSequenceSplitEligibility;
const splitTimelineSequenceFromSource = ({ nodePathInfo, splitFrame, }) => {
    const nodePath = nodePathInfo.sequenceSubscriptionKey;
    return (0, call_api_1.callApi)('/api/split-jsx-sequence', {
        fileName: nodePath.absolutePath,
        nodePath: nodePath.nodePath,
        splitFrame,
    })
        .then((result) => {
        if (result.success) {
            (0, NotificationCenter_1.showNotification)('Split sequence', 2000);
            return true;
        }
        (0, NotificationCenter_1.showNotification)(result.reason, 4000);
        return false;
    })
        .catch((err) => {
        (0, NotificationCenter_1.showNotification)(err.message, 4000);
        return false;
    });
};
exports.splitTimelineSequenceFromSource = splitTimelineSequenceFromSource;
const shouldHandleTimelineDuplicateShortcut = ({ shiftKey, }) => !shiftKey;
exports.shouldHandleTimelineDuplicateShortcut = shouldHandleTimelineDuplicateShortcut;
const shouldHandleTimelineSplitShortcut = ({ shiftKey, }) => shiftKey;
exports.shouldHandleTimelineSplitShortcut = shouldHandleTimelineSplitShortcut;
const splitSelectedTimelineItems = ({ selections, sequences, overrideIdsToNodePaths, propStatuses, splitFrame, splitSequence = exports.splitTimelineSequenceFromSource, }) => {
    var _a;
    if (selections.length !== 1) {
        return null;
    }
    const [selection] = selections;
    if (selection.type !== 'sequence') {
        return null;
    }
    const track = (0, find_track_for_node_path_info_1.findTrackForNodePathInfo)({
        sequences,
        overrideIdsToNodePaths,
        nodePathInfo: selection.nodePathInfo,
    });
    const sequencePropStatuses = propStatuses
        ? remotion_1.Internals.getPropStatusesCtx(propStatuses, selection.nodePathInfo.sequenceSubscriptionKey)
        : undefined;
    const eligibility = (0, exports.getTimelineSequenceSplitEligibility)({
        selection,
        sequence: (_a = track === null || track === void 0 ? void 0 : track.sequence) !== null && _a !== void 0 ? _a : null,
        splitFrame,
        propStatuses: sequencePropStatuses,
    });
    if (!eligibility.canSplit) {
        (0, NotificationCenter_1.showNotification)(eligibility.reason, 4000);
        return Promise.resolve(false);
    }
    return splitSequence({
        nodePathInfo: eligibility.nodePathInfo,
        splitFrame,
    });
};
exports.splitSelectedTimelineItems = splitSelectedTimelineItems;
