// @remotion/studio/dist/components/Timeline/TimelineSchemaField.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineFieldValue = exports.TimelineNonEditableStatus = exports.UnsupportedStatus = exports.TIMELINE_COMPUTED_EFFECT_FIX_LINK = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importDefault(require("react"));
const colors_1 = require("../../helpers/colors");
const get_timeline_keyframes_1 = require("./get-timeline-keyframes");
const TimelineArrayField_1 = require("./TimelineArrayField");
const TimelinePrimitiveFieldValue_1 = require("./TimelinePrimitiveFieldValue");
const INTERACTIVITY_BEST_PRACTICES_DOCS = 'https://www.remotion.dev/docs/studio/interactivity-best-practices';
const TIMELINE_COMPUTED_VALUE_FIX_LINK = `${INTERACTIVITY_BEST_PRACTICES_DOCS}#keep-editable-values-visible`;
exports.TIMELINE_COMPUTED_EFFECT_FIX_LINK = `${INTERACTIVITY_BEST_PRACTICES_DOCS}#keep-effects-editable`;
const unsupportedStatusWrapper = {
    alignItems: 'center',
    display: 'inline-flex',
    gap: 4,
    userSelect: 'none',
    WebkitUserSelect: 'none',
};
const unsupportedLabel = {
    color: colors_1.WHITE_ALPHA_40,
    fontSize: 12,
    fontStyle: 'italic',
};
const fixLinkBase = {
    color: colors_1.LIGHT_TEXT,
    display: 'inline-block',
    fontSize: 10,
    fontStyle: 'normal',
    fontWeight: 600,
    lineHeight: 1,
    textDecoration: 'none',
    width: 17,
};
const UnsupportedStatus = ({ label, fixHref }) => {
    const [hovered, setHovered] = react_1.default.useState(false);
    const [focused, setFocused] = react_1.default.useState(false);
    const visible = hovered || focused;
    const fixLink = react_1.default.useMemo(() => {
        return {
            ...fixLinkBase,
            opacity: visible ? 1 : 0,
            pointerEvents: visible ? 'auto' : 'none',
        };
    }, [visible]);
    const stopMousePropagation = (event) => {
        event.stopPropagation();
    };
    const stopPointerPropagation = (event) => {
        event.stopPropagation();
    };
    return (jsx_runtime_1.jsxs("span", { style: unsupportedStatusWrapper, onPointerEnter: () => setHovered(true), onPointerLeave: () => setHovered(false), children: [
            jsx_runtime_1.jsx("span", { style: unsupportedLabel, children: label }), fixHref ? (jsx_runtime_1.jsx("a", { href: fixHref, target: "_blank", rel: "noreferrer", style: fixLink, title: "Open docs to fix computed Studio values", onClick: stopMousePropagation, onDoubleClick: stopMousePropagation, onPointerDown: stopPointerPropagation, onFocus: () => setFocused(true), onBlur: () => setFocused(false), tabIndex: visible ? 0 : -1, children: "Fix" })) : null] }));
};
exports.UnsupportedStatus = UnsupportedStatus;
const TimelineNonEditableStatus = ({ propStatus }) => {
    if (propStatus.status === 'computed') {
        return (jsx_runtime_1.jsx(exports.UnsupportedStatus, { label: (0, get_timeline_keyframes_1.getComputedStatusLabel)(propStatus), fixHref: TIMELINE_COMPUTED_VALUE_FIX_LINK }));
    }
};
exports.TimelineNonEditableStatus = TimelineNonEditableStatus;
const TimelineFieldValue = ({ field, onSave, onDragValueChange, onDragEnd, propStatus, effectiveValue, scaleLockNodePath, }) => {
    if ((0, TimelinePrimitiveFieldValue_1.isTimelinePrimitiveFieldInfo)(field)) {
        return (jsx_runtime_1.jsx(TimelinePrimitiveFieldValue_1.TimelinePrimitiveFieldValue, { effectiveValue: effectiveValue, field: field, onDragEnd: onDragEnd, onDragValueChange: onDragValueChange, onSave: onSave, propStatus: propStatus, scaleLockNodePath: scaleLockNodePath }));
    }
    if (field.typeName === 'array') {
        return (jsx_runtime_1.jsx("span", { children: jsx_runtime_1.jsx(TimelineArrayField_1.TimelineArrayField, { effectiveValue: effectiveValue, field: field, onDragEnd: onDragEnd, onDragValueChange: onDragValueChange, onSave: onSave }) }));
    }
    throw new Error(`Unsupported field type: ${field.typeName}`);
};
exports.TimelineFieldValue = TimelineFieldValue;
