// @remotion/studio/dist/components/Timeline/TimelineTextContentField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineTextContentField = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const RemTextarea_1 = require("../NewComposition/RemTextarea");
const focus_inspector_field_1 = require("./focus-inspector-field");
const TimelineTextContentField = ({ effectiveValue, field, nodePath, onDragEnd, onDragValueChange, onSave, propStatus, }) => {
    var _a;
    const inputRef = (0, react_1.useRef)(null);
    const currentValue = String(effectiveValue !== null && effectiveValue !== void 0 ? effectiveValue : '');
    const draftRef = (0, react_1.useRef)({
        dirty: false,
        value: currentValue,
    });
    const latestRef = (0, react_1.useRef)({
        codeValue: propStatus.codeValue,
        onDragEnd,
        onDragValueChange,
        onSave,
    });
    latestRef.current = {
        codeValue: propStatus.codeValue,
        onDragEnd,
        onDragValueChange,
        onSave,
    };
    if (!draftRef.current.dirty) {
        draftRef.current.value = currentValue;
    }
    const commitPending = (0, react_1.useCallback)(() => {
        var _a;
        var _b, _c;
        if (!draftRef.current.dirty) {
            return false;
        }
        const value = (_b = (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.value) !== null && _b !== void 0 ? _b : draftRef.current.value;
        const savedValue = String((_c = latestRef.current.codeValue) !== null && _c !== void 0 ? _c : '');
        draftRef.current = {
            dirty: false,
            value,
        };
        if (value === savedValue) {
            latestRef.current.onDragEnd();
            return false;
        }
        latestRef.current
            .onSave(value)
            .finally(() => {
            if (!draftRef.current.dirty && draftRef.current.value === value) {
                latestRef.current.onDragEnd();
            }
        })
            .catch(() => undefined);
        return true;
    }, []);
    const setInputRef = (0, react_1.useCallback)((element) => {
        if (element === null && inputRef.current !== null) {
            commitPending();
        }
        inputRef.current = element;
        (0, focus_inspector_field_1.registerFocusInspectorFieldElement)({
            element,
            fieldKey: field.key,
            nodePath,
        });
        (0, focus_inspector_field_1.registerCommitPendingInspectorField)({
            commitPending: element === null ? null : commitPending,
            fieldKey: field.key,
            nodePath,
        });
    }, [commitPending, field.key, nodePath]);
    const onChange = (0, react_1.useCallback)((event) => {
        const { value } = event.currentTarget;
        draftRef.current = {
            dirty: true,
            value,
        };
        latestRef.current.onDragValueChange(value);
    }, []);
    const onKeyDownCapture = (0, react_1.useCallback)((event) => {
        var _a;
        if (event.key === 'Escape') {
            const savedValue = String((_a = latestRef.current.codeValue) !== null && _a !== void 0 ? _a : '');
            draftRef.current = {
                dirty: false,
                value: savedValue,
            };
            event.currentTarget.value = savedValue;
            latestRef.current.onDragEnd();
            event.currentTarget.blur();
        }
    }, []);
    return (jsx_runtime_1.jsx(RemTextarea_1.RemTextarea, { ref: setInputRef, status: "ok", small: true, defaultValue: currentValue, onBlur: commitPending, onChange: onChange, onKeyDownCapture: onKeyDownCapture, style: {
            boxSizing: 'border-box',
            height: 40,
            width: 160,
        } }, String((_a = propStatus.codeValue) !== null && _a !== void 0 ? _a : '')));
};
exports.TimelineTextContentField = TimelineTextContentField;
