// @remotion/studio/dist/components/Timeline/use-rename-sequence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRenameSequence = exports.getSequenceNameSaveAction = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const save_sequence_prop_1 = require("./save-sequence-prop");
const getSequenceNameSaveAction = ({ editedName, currentDisplayName, fallbackDisplayName, hasStaticNameProp, }) => {
    if (!hasStaticNameProp &&
        (editedName === '' || editedName === fallbackDisplayName)) {
        return { type: 'noop' };
    }
    if (hasStaticNameProp && editedName === currentDisplayName) {
        return { type: 'noop' };
    }
    return {
        type: 'save',
        defaultValue: editedName === '' ? JSON.stringify('') : null,
    };
};
exports.getSequenceNameSaveAction = getSequenceNameSaveAction;
const useRenameSequence = ({ clientId, nodePathInfo, sequence, validatedLocation, }) => {
    var _a;
    var _b, _c;
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const nodePath = (_b = nodePathInfo === null || nodePathInfo === void 0 ? void 0 : nodePathInfo.sequenceSubscriptionKey) !== null && _b !== void 0 ? _b : null;
    const propStatusesForOverride = (0, react_1.useMemo)(() => {
        return nodePath
            ? remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)
            : undefined;
    }, [propStatuses, nodePath]);
    const codeNameStatus = propStatusesForOverride === null || propStatusesForOverride === void 0 ? void 0 : propStatusesForOverride.name;
    const fallbackDisplayName = (_c = (_a = sequence.controls) === null || _a === void 0 ? void 0 : _a.componentName) !== null && _c !== void 0 ? _c : '<Sequence>';
    const staticNamePropValue = (codeNameStatus === null || codeNameStatus === void 0 ? void 0 : codeNameStatus.status) === 'static' &&
        typeof codeNameStatus.codeValue === 'string'
        ? codeNameStatus.codeValue
        : null;
    const hasStaticNameProp = staticNamePropValue !== null;
    const displayName = (0, react_1.useMemo)(() => {
        if (staticNamePropValue !== null) {
            return staticNamePropValue;
        }
        if ((codeNameStatus === null || codeNameStatus === void 0 ? void 0 : codeNameStatus.status) === 'static') {
            return fallbackDisplayName;
        }
        return sequence.displayName;
    }, [
        codeNameStatus === null || codeNameStatus === void 0 ? void 0 : codeNameStatus.status,
        fallbackDisplayName,
        sequence.displayName,
        staticNamePropValue,
    ]);
    const canRename = clientId !== null &&
        !window.remotion_isReadOnlyStudio &&
        Boolean(sequence.controls) &&
        nodePath !== null &&
        validatedLocation !== null &&
        codeNameStatus !== undefined &&
        codeNameStatus !== null &&
        codeNameStatus.status === 'static';
    const saveName = (0, react_1.useCallback)(async (name) => {
        if (!canRename ||
            clientId === null ||
            !sequence.controls ||
            !nodePath ||
            !validatedLocation) {
            return;
        }
        const action = (0, exports.getSequenceNameSaveAction)({
            editedName: name,
            currentDisplayName: displayName,
            fallbackDisplayName,
            hasStaticNameProp,
        });
        if (action.type === 'noop') {
            return;
        }
        await (0, save_sequence_prop_1.saveSequenceProps)({
            changes: [
                {
                    fileName: validatedLocation.source,
                    nodePath,
                    fieldKey: 'name',
                    value: name,
                    defaultValue: action.defaultValue,
                    schema: sequence.controls.schema,
                },
            ],
            setPropStatuses,
            clientId,
            undoLabel: 'Rename sequence',
            redoLabel: 'Rename sequence again',
        });
    }, [
        canRename,
        clientId,
        displayName,
        fallbackDisplayName,
        hasStaticNameProp,
        nodePath,
        sequence.controls,
        setPropStatuses,
        validatedLocation,
    ]);
    return {
        canRename,
        codeNameStatus,
        displayName,
        fallbackDisplayName,
        hasStaticNameProp,
        propStatusesForOverride,
        saveName,
    };
};
exports.useRenameSequence = useRenameSequence;
