// @remotion/studio/dist/components/Notifications/Notification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Notification = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../../helpers/colors");
const notification = {
    backgroundColor: colors_1.NOTIFICATION_BACKGROUND,
    color: colors_1.WHITE_HEX,
    fontFamily: 'Arial, Helvetica, sans-serif',
    display: 'inline-flex',
    padding: '6px 14px',
    borderRadius: 6,
    fontSize: 13,
    border: colors_1.NOTIFICATION_BORDER,
    boxShadow: colors_1.NOTIFICATION_SHADOW,
    marginTop: 3,
    marginBottom: 3,
    alignItems: 'center',
};
const Notification = ({ children, id, duration, created, onRemove }) => {
    (0, react_1.useEffect)(() => {
        if (duration === null) {
            return;
        }
        const timeout = setTimeout(() => {
            onRemove(id);
        }, duration - (Date.now() - created));
        return () => {
            clearTimeout(timeout);
        };
    }, [created, duration, id, onRemove]);
    return (jsx_runtime_1.jsx("div", { className: "css-reset", style: notification, children: children }));
};
exports.Notification = Notification;
