// @remotion/studio/dist/helpers/colors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BLUE_DISABLED = exports.BLUE_HOVERED = exports.BLUE = exports.WARNING_COLOR = exports.FAIL_COLOR = exports.BLACK_OPAQUE = exports.BLACK_ALPHA_90 = exports.BLACK_ALPHA_85 = exports.BLACK_ALPHA_80 = exports.BLACK_ALPHA_60 = exports.BLACK_ALPHA_50 = exports.BLACK_ALPHA_40 = exports.BLACK_ALPHA_30 = exports.BLACK_ALPHA_28 = exports.BLACK_ALPHA_10 = exports.WHITE_ALPHA_80 = exports.WHITE_ALPHA_72 = exports.WHITE_ALPHA_70 = exports.WHITE_ALPHA_60 = exports.WHITE_ALPHA_50 = exports.WHITE_ALPHA_45 = exports.WHITE_ALPHA_40 = exports.WHITE_ALPHA_35 = exports.WHITE_ALPHA_30 = exports.WHITE_ALPHA_25 = exports.WHITE_ALPHA_20 = exports.WHITE_ALPHA_15 = exports.WHITE_ALPHA_12 = exports.WHITE_ALPHA_10 = exports.WHITE_ALPHA_08 = exports.WHITE_ALPHA_06 = exports.WHITE_ALPHA_05 = exports.RULER_COLOR = exports.LIGHT_TEXT = exports.SELECTED_BACKGROUND = exports.LIGHT_COLOR = exports.INPUT_BACKGROUND = exports.BACKGROUND__TRANSPARENT = exports.BACKGROUND_HEX = exports.BACKGROUND = exports.BLACK_FULL_HEX = exports.BLACK_HEX = exports.WHITE_FULL_HEX = exports.WHITE_HEX = exports.CURRENT_COLOR_LOWERCASE = exports.CURRENT_COLOR = exports.TRANSPARENT = exports.RED = exports.WHITE = exports.BLACK = void 0;
exports.FOCUS_BOX_SHADOW = exports.HOT_MIDDLEWARE_WARNING_STYLE = exports.HOT_MIDDLEWARE_ERROR_STYLE = exports.NOTIFICATION_BORDER = exports.COLOR_PICKER_HANDLE_SHADOW = exports.COLOR_PICKER_POPUP_SHADOW = exports.SERVER_DISCONNECTED_SHADOW = exports.NOTIFICATION_SHADOW = exports.SHADOW_BLACK_ALPHA_50_TOWARDS_TOP = exports.SHADOW_BLACK_ALPHA_50_TOWARDS_BOTTOM = exports.SHADOW_BLACK = exports.BORDER_TIMELINE_MARQUEE_BLUE = exports.BORDER_TIMELINE_DROP_BLUE = exports.BORDER_STACK_FRAME_BLUE = exports.BORDER_INFO_BLUE = exports.BORDER_WHITE_ALPHA_20 = exports.BORDER_WHITE_ALPHA_12 = exports.BORDER_BLACK_ALPHA_60 = exports.BORDER_BLACK_ALPHA_50 = exports.BORDER_TRANSPARENT_2PX = exports.BORDER_CURRENT_COLOR = exports.BORDER_WHITE_2PX = exports.BORDER_WHITE = exports.BORDER_BLACK = exports.INFO_BLUE_BACKGROUND = exports.EASING_SELECTED_BACKGROUND = exports.TIMELINE_MARQUEE_BLUE_ALPHA_75 = exports.TIMELINE_MARQUEE_BLUE_ALPHA_16 = exports.TIMELINE_DROP_BLUE_ALPHA_75 = exports.TIMELINE_DROP_BLUE_ALPHA_16 = exports.TIMELINE_DROP_BLUE_ALPHA_12 = exports.TIMELINE_PLAYHEAD_COLOR = exports.TIMELINE_BLUE = exports.TIMELINE_SELECTED_LABEL_BACKGROUND_COLOR = exports.TIMELINE_SELECTED_BACKGROUND_COLOR = exports.TIMELINE_BACKGROUND_COLOR = exports.STACK_FRAME_BORDER_BLUE = exports.SERVER_DISCONNECTED_BACKGROUND = exports.INFO_BLUE = exports.ERROR_LINK_COLOR = exports.ERROR_CODE_FRAME_LINE_BACKGROUND = exports.ERROR_CODE_FRAME_BACKGROUND = exports.RENDER_STATUS_BACKGROUND = exports.NOTIFICATION_BACKGROUND = exports.KEYBOARD_SHORTCUT_KEY_COLOR = exports.LIGHT_GRAY = exports.TIMELINE_TRACK_SEPARATOR = exports.LINE_COLOR = exports.SELECTED_GUIDE = exports.UNSELECTED_GUIDE = void 0;
exports.getBackgroundFromHoverState = exports.TIMELINE_IMAGE_GRADIENT = exports.TIMELINE_VIDEO_GRADIENT = exports.TIMELINE_AUDIO_GRADIENT = exports.LOOPED_INDICATOR_DROP_SHADOW = exports.SELECTED_OUTLINE_UV_DROP_SHADOW = exports.SELECTED_OUTLINE_DROP_SHADOW = exports.ERROR_MESSAGE_MASK_IMAGE = exports.COLOR_PICKER_HUE_GRADIENT = exports.COLOR_PICKER_ALPHA_TRANSPARENT = exports.COLOR_PICKER_SATURATION_BLACK_GRADIENT = exports.COLOR_PICKER_SATURATION_VALUE_GRADIENT = exports.COLOR_PICKER_CHECKER_BACKGROUND_COLOR = exports.COLOR_PICKER_CHECKER_BACKGROUND_IMAGE = exports.CHECKERBOARD_BACKGROUND_IMAGE = void 0;
exports.BLACK = 'black';
exports.WHITE = 'white';
exports.RED = 'red';
exports.TRANSPARENT = 'transparent';
exports.CURRENT_COLOR = 'currentColor';
exports.CURRENT_COLOR_LOWERCASE = 'currentcolor';
exports.WHITE_HEX = '#fff';
exports.WHITE_FULL_HEX = '#ffffff';
exports.BLACK_HEX = '#000';
exports.BLACK_FULL_HEX = '#000000';
exports.BACKGROUND = 'rgb(31,36,40)';
exports.BACKGROUND_HEX = '#1f2428';
exports.BACKGROUND__TRANSPARENT = 'rgba(31,36,40, 0)';
exports.INPUT_BACKGROUND = '#2f363d';
exports.LIGHT_COLOR = '#ddd';
exports.SELECTED_BACKGROUND = 'hsla(0, 0%, 100%, 0.15)';
exports.LIGHT_TEXT = '#A6A7A9';
exports.RULER_COLOR = '#808080';
exports.WHITE_ALPHA_05 = 'rgba(255, 255, 255, 0.05)';
exports.WHITE_ALPHA_06 = 'rgba(255, 255, 255, 0.06)';
exports.WHITE_ALPHA_08 = 'rgba(255, 255, 255, 0.08)';
exports.WHITE_ALPHA_10 = 'rgba(255, 255, 255, 0.1)';
exports.WHITE_ALPHA_12 = 'rgba(255, 255, 255, 0.12)';
exports.WHITE_ALPHA_15 = 'rgba(255, 255, 255, 0.15)';
exports.WHITE_ALPHA_20 = 'rgba(255, 255, 255, 0.2)';
exports.WHITE_ALPHA_25 = 'rgba(255, 255, 255, 0.25)';
exports.WHITE_ALPHA_30 = 'rgba(255, 255, 255, 0.3)';
exports.WHITE_ALPHA_35 = 'rgba(255, 255, 255, 0.35)';
exports.WHITE_ALPHA_40 = 'rgba(255, 255, 255, 0.4)';
exports.WHITE_ALPHA_45 = 'rgba(255, 255, 255, 0.45)';
exports.WHITE_ALPHA_50 = 'rgba(255, 255, 255, 0.5)';
exports.WHITE_ALPHA_60 = 'rgba(255, 255, 255, 0.6)';
exports.WHITE_ALPHA_70 = 'rgba(255, 255, 255, 0.7)';
exports.WHITE_ALPHA_72 = 'rgba(255, 255, 255, 0.72)';
exports.WHITE_ALPHA_80 = 'rgba(255, 255, 255, 0.8)';
exports.BLACK_ALPHA_10 = 'rgba(0, 0, 0, 0.1)';
exports.BLACK_ALPHA_28 = 'rgba(0, 0, 0, 0.28)';
exports.BLACK_ALPHA_30 = 'rgba(0, 0, 0, 0.3)';
exports.BLACK_ALPHA_40 = 'rgba(0, 0, 0, 0.4)';
exports.BLACK_ALPHA_50 = 'rgba(0, 0, 0, 0.5)';
exports.BLACK_ALPHA_60 = 'rgba(0, 0, 0, 0.6)';
exports.BLACK_ALPHA_80 = 'rgba(0, 0, 0, 0.8)';
exports.BLACK_ALPHA_85 = 'rgba(0, 0, 0, 0.85)';
exports.BLACK_ALPHA_90 = 'rgba(0, 0, 0, 0.9)';
exports.BLACK_OPAQUE = 'rgba(0, 0, 0, 1)';
const SELECTED_HOVER_BACKGROUND = 'hsla(0, 0%, 100%, 0.25)';
exports.FAIL_COLOR = '#ff3232';
exports.WARNING_COLOR = '#f1c40f';
exports.BLUE = '#0b84f3';
exports.BLUE_HOVERED = '#4da3f7';
exports.BLUE_DISABLED = '#284f73';
exports.UNSELECTED_GUIDE = '#7e1219';
exports.SELECTED_GUIDE = '#d22d3a';
exports.LINE_COLOR = '#363A3E';
exports.TIMELINE_TRACK_SEPARATOR = '#13161B';
exports.LIGHT_GRAY = '#ccc';
exports.KEYBOARD_SHORTCUT_KEY_COLOR = '#eee';
exports.NOTIFICATION_BACKGROUND = '#111111';
exports.RENDER_STATUS_BACKGROUND = '#222';
exports.ERROR_CODE_FRAME_BACKGROUND = '#070707';
exports.ERROR_CODE_FRAME_LINE_BACKGROUND = '#121212';
exports.ERROR_LINK_COLOR = '#58a6ff';
exports.INFO_BLUE = '#60a5fa';
exports.SERVER_DISCONNECTED_BACKGROUND = '#e74c3c';
exports.STACK_FRAME_BORDER_BLUE = 'rgb(66, 144, 245)';
exports.TIMELINE_BACKGROUND_COLOR = '#0F1113';
exports.TIMELINE_SELECTED_BACKGROUND_COLOR = '#3B3F42';
exports.TIMELINE_SELECTED_LABEL_BACKGROUND_COLOR = '#B0B0B0';
exports.TIMELINE_BLUE = '#0b84ff';
exports.TIMELINE_PLAYHEAD_COLOR = '#f02c00';
exports.TIMELINE_DROP_BLUE_ALPHA_12 = 'rgba(0, 155, 255, 0.12)';
exports.TIMELINE_DROP_BLUE_ALPHA_16 = 'rgba(0, 155, 255, 0.16)';
exports.TIMELINE_DROP_BLUE_ALPHA_75 = 'rgba(0, 155, 255, 0.75)';
exports.TIMELINE_MARQUEE_BLUE_ALPHA_16 = 'rgba(70, 130, 255, 0.16)';
exports.TIMELINE_MARQUEE_BLUE_ALPHA_75 = 'rgba(70, 130, 255, 0.75)';
exports.EASING_SELECTED_BACKGROUND = 'rgba(11, 132, 243, 0.18)';
exports.INFO_BLUE_BACKGROUND = 'rgba(59, 130, 246, 0.15)';
exports.BORDER_BLACK = `1px solid ${exports.BLACK_HEX}`;
exports.BORDER_WHITE = `1px solid ${exports.WHITE_HEX}`;
exports.BORDER_WHITE_2PX = `2px solid ${exports.WHITE_HEX}`;
exports.BORDER_CURRENT_COLOR = `1px solid ${exports.CURRENT_COLOR}`;
exports.BORDER_TRANSPARENT_2PX = `2px solid ${exports.TRANSPARENT}`;
exports.BORDER_BLACK_ALPHA_50 = `1px solid ${exports.BLACK_ALPHA_50}`;
exports.BORDER_BLACK_ALPHA_60 = `1px solid ${exports.BLACK_ALPHA_60}`;
exports.BORDER_WHITE_ALPHA_12 = `1px solid ${exports.WHITE_ALPHA_12}`;
exports.BORDER_WHITE_ALPHA_20 = `1px solid ${exports.WHITE_ALPHA_20}`;
exports.BORDER_INFO_BLUE = '1px solid rgba(59, 130, 246, 0.4)';
exports.BORDER_STACK_FRAME_BLUE = `1px solid ${exports.STACK_FRAME_BORDER_BLUE}`;
exports.BORDER_TIMELINE_DROP_BLUE = `1px solid ${exports.TIMELINE_DROP_BLUE_ALPHA_75}`;
exports.BORDER_TIMELINE_MARQUEE_BLUE = `1px solid ${exports.TIMELINE_MARQUEE_BLUE_ALPHA_75}`;
exports.SHADOW_BLACK = `0 0 4px ${exports.BLACK}`;
exports.SHADOW_BLACK_ALPHA_50_TOWARDS_BOTTOM = `0 2px 8px ${exports.BLACK_ALPHA_50}`;
exports.SHADOW_BLACK_ALPHA_50_TOWARDS_TOP = `0 -2px 8px ${exports.BLACK_ALPHA_50}`;
exports.NOTIFICATION_SHADOW = `0 2px 3px ${exports.BLACK_OPAQUE}`;
exports.SERVER_DISCONNECTED_SHADOW = `0 2px 4px ${exports.BLACK_ALPHA_40}`;
exports.COLOR_PICKER_POPUP_SHADOW = `0 4px 16px ${exports.BLACK_ALPHA_50}`;
exports.COLOR_PICKER_HANDLE_SHADOW = `0 0 0 1px ${exports.BLACK_ALPHA_60}`;
exports.NOTIFICATION_BORDER = `0.25px solid ${exports.WHITE_ALPHA_10}`;
exports.HOT_MIDDLEWARE_ERROR_STYLE = 'color: #ff0000;';
exports.HOT_MIDDLEWARE_WARNING_STYLE = 'color: #999933;';
exports.FOCUS_BOX_SHADOW = `inset 1px 1px #555, inset -1px -1px #555, ` +
    `inset 1px -1px #555, inset -1px 1px #555`;
exports.CHECKERBOARD_BACKGROUND_IMAGE = `
     linear-gradient(
        45deg,
        ${exports.BLACK_ALPHA_10} 25%,
        ${exports.TRANSPARENT} 25%
      ),
      linear-gradient(135deg, ${exports.BLACK_ALPHA_10} 25%, ${exports.TRANSPARENT} 25%),
      linear-gradient(45deg, ${exports.TRANSPARENT} 75%, ${exports.BLACK_ALPHA_10} 75%),
      linear-gradient(135deg, ${exports.TRANSPARENT} 75%, ${exports.BLACK_ALPHA_10} 75%)
    `;
exports.COLOR_PICKER_CHECKER_BACKGROUND_IMAGE = `linear-gradient(45deg, #888 25%, ${exports.TRANSPARENT} 25%), ` +
    `linear-gradient(-45deg, #888 25%, ${exports.TRANSPARENT} 25%), ` +
    `linear-gradient(45deg, ${exports.TRANSPARENT} 75%, #888 75%), ` +
    `linear-gradient(-45deg, ${exports.TRANSPARENT} 75%, #888 75%)`;
exports.COLOR_PICKER_CHECKER_BACKGROUND_COLOR = '#444';
exports.COLOR_PICKER_SATURATION_VALUE_GRADIENT = `linear-gradient(to right, ${exports.WHITE_HEX}, ${exports.TRANSPARENT})`;
exports.COLOR_PICKER_SATURATION_BLACK_GRADIENT = `linear-gradient(to top, ${exports.BLACK_HEX}, ${exports.TRANSPARENT})`;
exports.COLOR_PICKER_ALPHA_TRANSPARENT = 'rgba(0, 0, 0, 0)';
exports.COLOR_PICKER_HUE_GRADIENT = 'linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)';
exports.ERROR_MESSAGE_MASK_IMAGE = `linear-gradient(to bottom, ${exports.WHITE} 60%, ${exports.TRANSPARENT})`;
exports.SELECTED_OUTLINE_DROP_SHADOW = `drop-shadow(0 0 1px ${exports.WHITE_ALPHA_20})`;
exports.SELECTED_OUTLINE_UV_DROP_SHADOW = `drop-shadow(0 1px 2px ${exports.BLACK_ALPHA_28})`;
exports.LOOPED_INDICATOR_DROP_SHADOW = `drop-shadow(0 0 2px ${exports.BLACK_ALPHA_90}) ` +
    `drop-shadow(0 1px 2px ${exports.BLACK_ALPHA_80})`;
exports.TIMELINE_AUDIO_GRADIENT = 'linear-gradient(rgb(16 171 58), rgb(43 165 63) 60%)';
exports.TIMELINE_VIDEO_GRADIENT = 'linear-gradient(to top, #8e44ad, #9b59b6)';
exports.TIMELINE_IMAGE_GRADIENT = 'linear-gradient(to top, #2980b9, #3498db)';
const getBackgroundFromHoverState = ({ selected, hovered, }) => {
    if (selected) {
        if (hovered) {
            return SELECTED_HOVER_BACKGROUND;
        }
        return exports.SELECTED_BACKGROUND;
    }
    if (hovered) {
        return exports.WHITE_ALPHA_06;
    }
    return exports.TRANSPARENT;
};
exports.getBackgroundFromHoverState = getBackgroundFromHoverState;
