// @remotion/studio/dist/helpers/use-rename-composition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRenameComposition = void 0;
const react_1 = require("react");
const actions_1 = require("../components/RenderQueue/actions");
const url_state_1 = require("./url-state");
const validate_new_comp_data_1 = require("./validate-new-comp-data");
const useRenameComposition = ({ compositions, currentId, newId, }) => {
    const getValidationMessage = (0, react_1.useCallback)((value) => {
        if (value === currentId) {
            return null;
        }
        return (0, validate_new_comp_data_1.validateCompositionName)(value, compositions);
    }, [compositions, currentId]);
    const getCodemod = (0, react_1.useCallback)((value) => {
        return {
            type: 'rename-composition',
            idToRename: currentId,
            newId: value,
        };
    }, [currentId]);
    const validationMessage = (0, react_1.useMemo)(() => {
        return getValidationMessage(newId);
    }, [getValidationMessage, newId]);
    const codemod = (0, react_1.useMemo)(() => {
        return getCodemod(newId);
    }, [getCodemod, newId]);
    const valid = validationMessage === null && currentId !== newId;
    const renameComposition = (0, react_1.useCallback)(async ({ newCompositionId, signal, symbolicatedStack, }) => {
        const result = await (0, actions_1.applyCodemod)({
            codemod: getCodemod(newCompositionId),
            dryRun: false,
            signal,
            symbolicatedStack,
        });
        if (result.success) {
            (0, url_state_1.pushUrl)(`/${newCompositionId}`);
        }
        return result;
    }, [getCodemod]);
    return {
        codemod,
        getValidationMessage,
        renameComposition,
        valid,
        validationMessage,
    };
};
exports.useRenameComposition = useRenameComposition;
