// @remotion/studio/dist/helpers/use-create-composition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCreateComposition = exports.getUniqueCompositionName = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const actions_1 = require("../components/RenderQueue/actions");
const url_state_1 = require("./url-state");
const validate_new_comp_data_1 = require("./validate-new-comp-data");
const toPascalCase = (value) => {
    var _a;
    const words = (_a = value.match(/[a-zA-Z0-9]+/g)) !== null && _a !== void 0 ? _a : [];
    const candidate = words
        .map((word) => `${word.charAt(0).toUpperCase()}${word.slice(1)}`)
        .join('');
    if (!candidate) {
        return 'NewComposition';
    }
    if (/^[0-9]/.test(candidate)) {
        return `Composition${candidate}`;
    }
    return candidate;
};
const waitForComposition = (compositionId) => {
    return new Promise((resolve) => {
        const started = Date.now();
        const interval = window.setInterval(() => {
            var _a;
            var _b;
            const compositionNames = (_b = (_a = window.remotion_getCompositionNames) === null || _a === void 0 ? void 0 : _a.call(window)) !== null && _b !== void 0 ? _b : [];
            if (compositionNames.includes(compositionId) ||
                Date.now() - started > 10000) {
                window.clearInterval(interval);
                resolve();
            }
        }, 100);
    });
};
const getUniqueCompositionName = (compositions) => {
    let counter = 1;
    while (true) {
        const name = counter === 1 ? 'NewComposition' : `NewComposition${counter}`;
        const err = (0, validate_new_comp_data_1.validateCompositionName)(name, compositions);
        if (!err) {
            return name;
        }
        counter++;
    }
};
exports.getUniqueCompositionName = getUniqueCompositionName;
const useCreateComposition = ({ compositions, durationInFrames, folderName, newId, parentName, selectedFrameRate, size, }) => {
    const { setCanvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionSetters);
    const componentName = (0, react_1.useMemo)(() => toPascalCase(newId), [newId]);
    const nameValidationMessage = (0, react_1.useMemo)(() => {
        return (0, validate_new_comp_data_1.validateCompositionName)(newId, compositions);
    }, [compositions, newId]);
    const widthValidationMessage = (0, react_1.useMemo)(() => {
        return (0, validate_new_comp_data_1.validateCompositionDimension)('Width', size.width);
    }, [size.width]);
    const heightValidationMessage = (0, react_1.useMemo)(() => {
        return (0, validate_new_comp_data_1.validateCompositionDimension)('Height', size.height);
    }, [size.height]);
    const codemod = (0, react_1.useMemo)(() => {
        return {
            type: 'new-composition',
            newDurationInFrames: Number(durationInFrames),
            newFps: Number(selectedFrameRate),
            newHeight: Number(size.height),
            newWidth: Number(size.width),
            newId,
            componentName,
            componentImportPath: `./${componentName}`,
            folderName,
            parentName,
        };
    }, [
        componentName,
        durationInFrames,
        folderName,
        newId,
        parentName,
        selectedFrameRate,
        size.height,
        size.width,
    ]);
    const valid = nameValidationMessage === null &&
        widthValidationMessage === null &&
        heightValidationMessage === null;
    const createComposition = (0, react_1.useCallback)(async ({ signal, symbolicatedStack, }) => {
        const result = await (0, actions_1.applyCodemod)({
            codemod,
            dryRun: false,
            signal,
            symbolicatedStack,
        });
        if (result.success) {
            await waitForComposition(newId);
            setCanvasContent({ type: 'composition', compositionId: newId });
            (0, url_state_1.pushUrl)(`/${newId}`);
        }
        return result;
    }, [codemod, newId, setCanvasContent]);
    return {
        codemod,
        createComposition,
        heightValidationMessage,
        nameValidationMessage,
        valid,
        widthValidationMessage,
    };
};
exports.useCreateComposition = useCreateComposition;
