// @remotion/studio-shared/dist/composition-drag-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseCompositionDragData = exports.makeCompositionDragData = exports.COMPOSITION_DRAG_MIME_TYPE = void 0;
exports.COMPOSITION_DRAG_MIME_TYPE = 'application/vnd.remotion.composition+json';
const isRecord = (value) => {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
};
const isCompositionId = (value) => {
    return (typeof value === 'string' &&
        value.length > 0 &&
        value.length < 500 &&
        /^([a-zA-Z0-9-\u4E00-\u9FFF])+$/.test(value));
};
const hasParentDirectorySegment = (value) => {
    return value.split('/').includes('..');
};
const isCompositionFile = (value) => {
    if (value === null) {
        return true;
    }
    return (typeof value === 'string' &&
        value.length > 0 &&
        value.length < 2000 &&
        !value.includes('\0') &&
        !value.includes('\\') &&
        !value.startsWith('/') &&
        !hasParentDirectorySegment(value));
};
const makeCompositionDragData = ({ compositionFile, compositionId, }) => {
    return {
        type: 'remotion-composition',
        version: 1,
        compositionFile,
        compositionId,
    };
};
exports.makeCompositionDragData = makeCompositionDragData;
const parseCompositionDragData = (value) => {
    try {
        const parsed = JSON.parse(value);
        if (!isRecord(parsed)) {
            return null;
        }
        if (parsed.type !== 'remotion-composition' || parsed.version !== 1) {
            return null;
        }
        if (!isCompositionId(parsed.compositionId) ||
            !isCompositionFile(parsed.compositionFile)) {
            return null;
        }
        return (0, exports.makeCompositionDragData)({
            compositionFile: parsed.compositionFile,
            compositionId: parsed.compositionId,
        });
    }
    catch (_a) {
        return null;
    }
};
exports.parseCompositionDragData = parseCompositionDragData;
