// @remotion/studio/dist/components/NewComposition/NewFolder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewFolder = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const get_folder_id_1 = require("../../helpers/get-folder-id");
const validate_new_folder_name_1 = require("../../helpers/validate-new-folder-name");
const layout_1 = require("../layout");
const ModalFooter_1 = require("../ModalFooter");
const ModalHeader_1 = require("../ModalHeader");
const layout_2 = require("../RenderModal/layout");
const actions_1 = require("../RenderQueue/actions");
const CodemodFooter_1 = require("./CodemodFooter");
const DismissableModal_1 = require("./DismissableModal");
const InputAndValidationContainer_1 = require("./InputAndValidationContainer");
const RemInput_1 = require("./RemInput");
const ValidationMessage_1 = require("./ValidationMessage");
const content = {
    padding: 12,
    paddingRight: 12,
    flex: 1,
    fontSize: 13,
    minWidth: 500,
};
const getUniqueFolderName = ({ folders, parentName, }) => {
    let counter = 1;
    while (true) {
        const name = counter === 1 ? 'NewFolder' : `NewFolder${counter}`;
        const err = (0, validate_new_folder_name_1.validateNewFolderName)({ folders, newName: name, parentName });
        if (!err) {
            return name;
        }
        counter++;
    }
};
const NewFolder = ({ parentName, stack }) => {
    const { folders } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const [newName, setName] = (0, react_1.useState)(() => getUniqueFolderName({ folders, parentName }));
    const inputRef = (0, react_1.useRef)(null);
    (0, react_1.useEffect)(() => {
        const input = inputRef.current;
        if (!input)
            return;
        input.select();
    }, []);
    const onNameChange = (0, react_1.useCallback)((e) => {
        setName(e.target.value);
    }, []);
    const folderNameErrMessage = (0, validate_new_folder_name_1.validateNewFolderName)({
        folders,
        newName,
        parentName,
    });
    const valid = folderNameErrMessage === null;
    const codemod = (0, react_1.useMemo)(() => {
        return {
            type: 'new-folder',
            folderName: newName,
            parentName,
        };
    }, [newName, parentName]);
    const onSubmit = (0, react_1.useCallback)((e) => {
        e.preventDefault();
    }, []);
    const folderId = (0, get_folder_id_1.getFolderId)({ folderName: newName, parentName });
    return (jsx_runtime_1.jsxs(DismissableModal_1.DismissableModal, { children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: "New folder" }), jsx_runtime_1.jsxs("form", { onSubmit: onSubmit, children: [
                    jsx_runtime_1.jsxs("div", { style: content, children: [parentName ? (jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                    jsx_runtime_1.jsx("div", { style: layout_2.label, children: "Parent" }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: parentName })
                                ] })) : null, jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                    jsx_runtime_1.jsx("div", { style: layout_2.label, children: "Name" }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsxs(InputAndValidationContainer_1.InputAndValidationContainer, { children: [
                                                jsx_runtime_1.jsx(RemInput_1.RemotionInput, { ref: inputRef, value: newName, onChange: onNameChange, type: "text", autoFocus: true, placeholder: "Folder name", status: "ok", rightAlign: true }), folderNameErrMessage ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                                                        jsx_runtime_1.jsx(layout_1.Spacing, { y: 1, block: true }), jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { align: "flex-start", message: folderNameErrMessage, type: "error" })
                                                    ] })) : null] }) })
                                ] })
                        ] }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsx(CodemodFooter_1.CodemodFooter, { loadingNotification: 'Creating folder...', errorNotification: 'Could not create folder', successNotification: `Created folder ${folderId}`, genericSubmitLabel: 'Add to root file', submitLabel: ({ relativeRootPath }) => `Add to ${relativeRootPath}`, codemod: codemod, stack: stack, valid: valid, onSuccess: null, fallbackToRootFile: true, applyCodemod: ({ signal, symbolicatedStack }) => (0, actions_1.applyCodemod)({
                                codemod,
                                dryRun: false,
                                signal,
                                symbolicatedStack,
                            }) }) })
                ] })
        ] }));
};
exports.NewFolder = NewFolder;
