// @remotion/studio/dist/components/NewComposition/RenameStaticFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenameStaticFileModal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const modals_1 = require("../../state/modals");
const Button_1 = require("../Button");
const layout_1 = require("../layout");
const ModalFooter_1 = require("../ModalFooter");
const ModalHeader_1 = require("../ModalHeader");
const layout_2 = require("../RenderModal/layout");
const use_static_files_1 = require("../use-static-files");
const DismissableModal_1 = require("./DismissableModal");
const InputAndValidationContainer_1 = require("./InputAndValidationContainer");
const RemInput_1 = require("./RemInput");
const use_rename_static_file_1 = require("./use-rename-static-file");
const ValidationMessage_1 = require("./ValidationMessage");
const content = {
    padding: 12,
    paddingRight: 12,
    flex: 1,
    fontSize: 13,
    minWidth: 500,
};
const RenameStaticFileModal = ({ relativePath }) => {
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const staticFiles = (0, use_static_files_1.useStaticFiles)();
    const [newName, setNewName] = (0, react_1.useState)(() => (0, use_rename_static_file_1.getStaticFileBaseName)(relativePath));
    const [submitting, setSubmitting] = (0, react_1.useState)(false);
    const inputRef = (0, react_1.useRef)(null);
    const renameFile = (0, use_rename_static_file_1.useRenameStaticFile)({ relativePath, staticFiles });
    (0, react_1.useEffect)(() => {
        const input = inputRef.current;
        if (!input)
            return;
        const [start, end] = (0, use_rename_static_file_1.getStaticFileRenameSelection)(newName);
        input.setSelectionRange(start, end);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    const newRelativePath = (0, react_1.useMemo)(() => {
        return (0, use_rename_static_file_1.getRenamedStaticFilePath)({ newName, relativePath });
    }, [newName, relativePath]);
    const changed = newRelativePath !== relativePath;
    const validationMessage = (0, react_1.useMemo)(() => {
        return (0, use_rename_static_file_1.validateStaticFileRename)({
            newName,
            newRelativePath,
            relativePath,
            staticFiles,
        });
    }, [newName, newRelativePath, relativePath, staticFiles]);
    const valid = changed && validationMessage === null;
    const onNameChange = (0, react_1.useCallback)((e) => {
        setNewName(e.target.value);
    }, []);
    const onRename = (0, react_1.useCallback)(() => {
        if (!valid) {
            return;
        }
        setSubmitting(true);
        renameFile(newName)
            .then((renamed) => {
            if (!renamed) {
                setSubmitting(false);
                return;
            }
            setSelectedModal(null);
        })
            .catch(() => {
            setSubmitting(false);
        });
    }, [newName, renameFile, setSelectedModal, valid]);
    const onSubmit = (0, react_1.useCallback)((e) => {
        e.preventDefault();
        onRename();
    }, [onRename]);
    return (jsx_runtime_1.jsxs(DismissableModal_1.DismissableModal, { children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: `Rename ${relativePath}` }), jsx_runtime_1.jsxs("form", { onSubmit: onSubmit, children: [
                    jsx_runtime_1.jsx("div", { style: content, children: jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                jsx_runtime_1.jsx("div", { style: layout_2.label, children: "Name" }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsxs(InputAndValidationContainer_1.InputAndValidationContainer, { children: [
                                            jsx_runtime_1.jsx(RemInput_1.RemotionInput, { ref: inputRef, value: newName, onChange: onNameChange, type: "text", autoFocus: true, autoComplete: "off", "data-1p-ignore": true, placeholder: "Asset name", status: validationMessage ? 'error' : 'ok', rightAlign: true }), validationMessage ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                                                    jsx_runtime_1.jsx(layout_1.Spacing, { y: 1, block: true }), jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { align: "flex-start", message: validationMessage, type: "error" })
                                                ] })) : null] }) })
                            ] }) }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsxs(layout_1.Row, { align: "center", justify: "flex-end", children: [
                                jsx_runtime_1.jsx(Button_1.Button, { onClick: () => setSelectedModal(null), children: "Cancel" }), jsx_runtime_1.jsx(layout_1.Spacing, { x: 1 }), jsx_runtime_1.jsx(Button_1.Button, { onClick: onRename, disabled: !valid || submitting, children: "Rename" })
                            ] }) })
                ] })
        ] }));
};
exports.RenameStaticFileModal = RenameStaticFileModal;
