// @remotion/studio/dist/components/NewComposition/use-rename-static-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRenameStaticFile = exports.validateStaticFileRename = exports.getRenamedStaticFilePath = exports.getStaticFileRenameSelection = exports.getStaticFileBaseName = exports.getStaticFileParent = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const rename_static_file_1 = require("../../api/rename-static-file");
const url_state_1 = require("../../helpers/url-state");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const getStaticFileParent = (relativePath) => {
    const slashIndex = relativePath.lastIndexOf('/');
    return slashIndex === -1 ? '' : relativePath.slice(0, slashIndex);
};
exports.getStaticFileParent = getStaticFileParent;
const getStaticFileBaseName = (relativePath) => {
    const slashIndex = relativePath.lastIndexOf('/');
    return slashIndex === -1 ? relativePath : relativePath.slice(slashIndex + 1);
};
exports.getStaticFileBaseName = getStaticFileBaseName;
const getStaticFileRenameSelection = (fileName) => {
    const dotIndex = fileName.lastIndexOf('.');
    return [0, dotIndex > 0 ? dotIndex : fileName.length];
};
exports.getStaticFileRenameSelection = getStaticFileRenameSelection;
const getRenamedStaticFilePath = ({ newName, relativePath, }) => {
    const parent = (0, exports.getStaticFileParent)(relativePath);
    return [parent, newName].filter(Boolean).join('/');
};
exports.getRenamedStaticFilePath = getRenamedStaticFilePath;
const validateStaticFileRename = ({ newName, newRelativePath, relativePath, staticFiles, }) => {
    const trimmedName = newName.trim();
    if (trimmedName.length === 0) {
        return 'Name cannot be empty';
    }
    if (trimmedName !== newName) {
        return 'Name cannot start or end with whitespace';
    }
    if (newName.includes('/') || newName.includes('\\')) {
        return 'Name cannot include slashes';
    }
    const existingFile = staticFiles.find((file) => {
        return file.name === newRelativePath && file.name !== relativePath;
    });
    if (existingFile) {
        return 'An asset with this name already exists';
    }
    return null;
};
exports.validateStaticFileRename = validateStaticFileRename;
const useRenameStaticFile = ({ relativePath, staticFiles, }) => {
    const { canvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const { setCanvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionSetters);
    return (0, react_1.useCallback)(async (newName) => {
        const newRelativePath = (0, exports.getRenamedStaticFilePath)({
            newName,
            relativePath,
        });
        if (newRelativePath === relativePath) {
            return true;
        }
        const validationMessage = (0, exports.validateStaticFileRename)({
            newName,
            newRelativePath,
            relativePath,
            staticFiles,
        });
        if (validationMessage) {
            (0, NotificationCenter_1.showNotification)(validationMessage, 2000);
            return false;
        }
        const notification = (0, NotificationCenter_1.showNotification)(`Renaming ${relativePath}...`, null);
        try {
            await (0, rename_static_file_1.renameStaticFile)({
                oldRelativePath: relativePath,
                newRelativePath,
            });
            if ((canvasContent === null || canvasContent === void 0 ? void 0 : canvasContent.type) === 'asset' &&
                canvasContent.asset === relativePath) {
                setCanvasContent({ type: 'asset', asset: newRelativePath });
                (0, url_state_1.pushUrl)(`/assets/${newRelativePath}`);
            }
            notification.replaceContent(`Renamed to ${newRelativePath}`, 2000);
            return true;
        }
        catch (err) {
            notification.replaceContent(`Could not rename ${relativePath}: ${err.message}`, 3000);
            return false;
        }
    }, [canvasContent, relativePath, setCanvasContent, staticFiles]);
};
exports.useRenameStaticFile = useRenameStaticFile;
