// @remotion/studio-server/dist/preview-server/routes/split-jsx-sequence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitJsxSequenceHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const split_jsx_sequence_1 = require("../../codemods/split-jsx-sequence");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const log_update_1 = require("./log-updates/log-update");
const source_file_write_queue_1 = require("./source-file-write-queue");
const splitJsxSequenceHandler = ({ input: { fileName, nodePath, splitFrame }, remotionRoot, logLevel }) => (0, source_file_write_queue_1.withSourceFileWriteQueue)(async () => {
    try {
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[split-jsx-sequence] Received request for fileName="${fileName}" at frame ${splitFrame}`);
        const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName,
            action: 'modify',
        });
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const { output, formatted, nodeLabel, logLine } = await (0, split_jsx_sequence_1.splitJsxSequence)({
            input: fileContents,
            nodePath,
            splitFrame,
        });
        (0, undo_stack_1.pushToUndoStack)({
            filePath: absolutePath,
            oldContents: fileContents,
            newContents: null,
            logLevel,
            remotionRoot,
            logLine,
            description: {
                undoMessage: `↩️  Split of ${nodeLabel}`,
                redoMessage: `↪️  Split of ${nodeLabel}`,
            },
            entryType: 'split-jsx-sequence',
            suppressHmrOnFileRestore: false,
        });
        (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, undefined);
        const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath,
            line: logLine,
        });
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} Split ${nodeLabel}`);
        if (!formatted) {
            (0, log_update_1.warnAboutPrettierOnce)(logLevel);
        }
        renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, `[split-jsx-sequence] Wrote ${fileRelativeToRoot}${formatted ? ' (formatted)' : ''}`);
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
});
exports.splitJsxSequenceHandler = splitJsxSequenceHandler;
